admin/tests/unit/controllers/edit-user.js

```diff
@@ -566,5 +567,210 @@ describe('EditUserCtrl controller', () => {
+        scope.editUserModel.token_login = true;
+        scope.editUserModel.phone = '+40755696969';
+
+        setTimeout(() => {
+          scope.editUser();
+          setTimeout(() => {
+            chai.expect(CreateUser.callCount).to.equal(1);
+            chai.expect(CreateUser.args[0][0]).to.deep.include({
+              username: 'newuser',
+              phone: '+40755696969',
+              roles: ['data-entry'],
+              token_login: true,
+            });
+            done();
+          }, 10);
+        });
+      });
+    });
+
+    it('should not overwrite token_login when editing and making no changes', (done) => {
+      Settings = sinon.stub().resolves({
+        roles: {
+          'district-manager': { name: 'xyz', offline: true }, 'data-entry': { name: 'abc' },
+          supervisor: { name: 'qrt', offline: true },
+          'national-manager': { name: 'national-manager', offline: false }
+        },
+        token_login: {
+          translation_key: 'key',
+          app_url: 'url',
+        }
+      });
+      http.get.withArgs('/api/v1/users-info').resolves({ data: { warn: false, total_docs: 100, limit: 10000 } });
+      Translate.fieldIsRequired.resolves('Facility field is a required field');
+
+      userToEdit.token_login = true;
+      mockEditAUser(userToEdit);
+      setTimeout(() => {
+        scope.editUserModel.phone = '+40755696969';
+        scope.editUserModel.role = 'data-entry';
+
+        setTimeout(() => {
+          scope.editUser();
+          setTimeout(() => {
+            chai.expect(UpdateUser.callCount).to.equal(1);
+            chai.expect(UpdateUser.args[0][1]).to.deep.include({
+              phone: '+40755696969',
+              roles: ['data-entry']
+            });
+            done();
+          }, 10);
+        });
+      });
+    });
+
+    it('should require password when disabling token_login', (done) => {
+      Settings = sinon.stub().resolves({
+        roles: {
+          'district-manager': { name: 'xyz', offline: true }, 'data-entry': { name: 'abc' },
+          supervisor: { name: 'qrt', offline: true },
+          'national-manager': { name: 'national-manager', offline: false }
+        },
+        token_login: {
+          translation_key: 'key',
+          app_url: 'url',
+        }
+      });
+      http.get.withArgs('/api/v1/users-info').resolves({ data: { warn: false, total_docs: 100, limit: 10000 } });
+      Translate.fieldIsRequired.withArgs('Password').resolves('password required');
+
+      userToEdit.token_login = true;
+      mockEditAUser(userToEdit);
+      setTimeout(() => {
+        scope.editUserModel.token_login = false;
+        scope.editUserModel.phone = '';
+        scope.editUserModel.role = 'data-entry';
+
+
+        setTimeout(() => {
+          scope.editUser();
+          setTimeout(() => {
+            chai.expect(UpdateUser.callCount).to.equal(0);
+            chai.expect(scope.errors.password).to.equal('password required');
```

Comment:
I _think_ this will throw an uncaught error if the expectation fails... do we need to wrap it in a try/catch and call `done(err)`? Same for all the tests in this suite.

Reply:
:+1: wrapped every timeout in try catch